avalanche/training/templates/online_supervised.py

```diff
@@ -1,209 +1,318 @@
+        other_dataloader_args = {}
 
-        # if self._stop_training:  # Early stopping
-        #     self._stop_training = False
-        # break
+        if parse_version(torch.__version__) >= parse_version("1.7.0"):
+            other_dataloader_args["persistent_workers"] = persistent_workers
 
-        for self.n_pass in range(self.num_passes):
-            self.training_epoch(**kwargs)
+        self.dataloader = DataLoader(
+            self.adapted_dataset,
+            num_workers=num_workers,
+            batch_size=self.eval_mb_size,
+            pin_memory=pin_memory,
+            **other_dataloader_args
+        )
 
-        # if is_last_sub_exp:
-        self._after_training_epoch(**kwargs)
-        self._after_training_exp(**kwargs)
+    def forward(self):
+        """Compute the model's output given the current mini-batch."""
+        return avalanche_forward(self.model, self.mb_x, self.mb_task_id)
 
     def model_adaptation(self, model=None):
-        """Adapts the model to the data from the current
-           (full) experience.
+        """Adapts the model to the current data.
 
         Calls the :class:`~avalanche.models.DynamicModule`s adaptation.
         """
         if model is None:
             model = self.model
 
-        for module in model.modules():
-            if isinstance(module, DynamicModule):
-                module.adaptation(self.full_experience)
+        # For training:
+        if isinstance(self.experience, OnlineCLExperience):
+            if self.experience.access_task_boundaries:
+                avalanche_model_adaptation(model,
+                                           self.experience.origin_experience)
+            else:
+                avalanche_model_adaptation(model, self.experience)
+        # For evaluation, the experience is not necessarily an online
+        # experience:
+        else:
+            avalanche_model_adaptation(model, self.experience)
+
         return model.to(self.device)
+
+    def _unpack_minibatch(self):
+        """We assume mini-batches have the form <x, y, ..., t>.
+        This allows for arbitrary tensors between y and t.
+        Keep in mind that in the most general case mb_task_id is a tensor
+        which may contain different labels for each sample.
+        """
+        assert len(self.mbatch) >= 3
+        super()._unpack_minibatch()
+
+    def eval_dataset_adaptation(self, **kwargs):
+        """Initialize `self.adapted_dataset`."""
+        self.adapted_dataset = self.experience.dataset
+        self.adapted_dataset = self.adapted_dataset.eval()
+
+    def make_optimizer(self):
+        """Optimizer initialization.
+
+        Called before each training experiene to configure the optimizer.
+        """
+        # we reset the optimizer's state after each experience.
+        # This allows to add new parameters (new heads) and
+        # freezing old units during the model's adaptation phase.
+        reset_optimizer(self.optimizer, self.model)
```

Comment:
is this correct? This would reset the momentum after each mini-batch. It's ok in CIL, not in OCL.